libraries/Wire/src/Wire.cpp

```diff
@@ -259,20 +293,23 @@ size_t TwoWire::write(uint8_t data)
   return 1;
 }
 
-// must be called in:
-// slave tx event callback
-// or after beginTransmission(address)
-size_t TwoWire::write(const uint8_t *data, size_t quantity)
+/**
+  * @brief  This function must be called in slave Tx event callback or after
+  *         beginTransmission() and before endTransmission().
+  * @param  pdata: pointer to the buffer data
+  * @param  quantity: number of bytes to write
+  * @retval number of bytes read. If > 0 then data are present in pdata buffer.
+  */
+size_t TwoWire::write(const uint8_t *pdata, size_t quantity)
 {
   if(transmitting){
   // in master transmitter mode
-    for(size_t i = 0; i < quantity; ++i){
-      write(data[i]);
-    }
+    txBufferLength = quantity;
+    pWriteData = (uint8_t*)pdata;
```

Comment:
before the patch, write() was called and transmission happens.
Now only 2 variables are updated - how does transmission actually start ?

Reply:
I think you confused master and slave mode. I only change the master mode.
The transmission only happens when endtransmission() is called (before and now).

Reply:
well what I see here is that you remove the lines:
-    for(size_t i = 0; i < quantity; ++i){
 -      write(data[i]);
 -    }
and replace them by txBufferLength and pWriteData assignment ...  it looks like changing the dynamic

Reply:
well what I see here is that you remove the lines:
for(size_t i = 0; i < quantity; ++i){
 write(data[i]);
}
and replace them by txBufferLength and pWriteData assignment ... it looks like changing the dynamic

Reply:
Of course I changed the dynamic because I pass directly the data buffer pWriteData to the endTransmission() function (and so to the HAL) and I don't use anymore the internal buffer (low size) which is filled in the write() function.

Reply:
Ok - I got confused by write() function as I thought it would actually transmit data on I2C lines. Thanks for the clarification.

Reply:
In fact, @LMESTM is right. the dynamic is not the same and so the txbuffer construction is broken.
I've seen the issue with test on i2C eeprom. See code below:
```
        Wire.beginTransmission(ctrlByte);
        if (_nAddrBytes == 2) Wire.write( (byte) (addr >> 8) );   //high addr byte
        Wire.write( (byte) addr );                                //low addr byte
        Wire.write(values, nWrite);
        txStatus = Wire.endTransmission();
```

The 2 first calls of `Wire.write()` put the bytes in the internal `txBuffer` but when `Wire.write(values, nWrite);` is called, it initialize `pWriteData`.
Finally when `Wire.endTransmission();` is called:
```
    // Select the buffer where are stored the data to send
    if(pWriteData != NULL) {
      data = pWriteData;
    } else {
      data = txBuffer;
    }
```
the `txBuffer` datas are ignored....

Reply:
There are 2 methods completely different now. You can't mix them as you see.

But maybe I found a solution to mix them and keep all older sketches compatibles.
I look at this.